text/0000-newtype.md

```diff
@@ -0,0 +1,360 @@
+
+fn f(v: New) { }
+f(1);
+
+fn g(v: u32) { }
+g(New(1));
+
+let x: New = 1;
+let y: u32 = New(1);
+
+let a = 1 as New;
+let b = New(1) as u32;
+```
+
+The reason for this is that, as completely new types, the semantic of the
+newtype structs can freely deviate from the semantics of the wrapped type:
+
+```rust
+struct New(u32);
+
+impl Drop for New {
+    fn drop(&mut self) { }
+}
+```
+
+Other examples include:
+
+- The wrapped type is `Send` but the newtype struct is not `Send`.
+- The wrapped type is not `Send` but the newtype struct is `Send`.
+
+## The problem
+
+Above we've already seen an example where neither of these alias types are
+appropriate:
+
+```rust
+#[cfg(unix)]
+type c_long = i64;
+#[cfg(windows)]
+type c_long = i32;
+```
+
+On 64 bit systems, these are the definitions of `c_long` in liblibc. Consider
+the following fictional C function which we are trying to wrap:
+
+```c
+int Gen(long input);
+```
+
+In Rust the signature looks like this:
+
+```rust
+extern {
+    fn Gen(input: c_long) -> c_int;
+}
+```
+
+We create a rustic wrapper:
+
+```rust
+fn gen(input: i64) -> i32 {
+    unsafe { Gen(input) }
+}
+```
+
+This will work fine on our Linux development box but on windows the compilation
+will fail because `c_long` is `i32`.
+
+Correct, platform independent code will have to add explicit casts:
+
+```rust
+fn gen(input: i64) -> i32 {
+    unsafe { Gen(input as c_long) as i32 }
+}
+```
+
+There are currently some proposals regarding `as` that would worsen this
+situation:
+
+1. Make useless casts an error or warning: `input as c_long` would cause an
+   error or warning because `input` and `c_long` are both `i64`.
+2. Add implicit widening: This would add more cases where `as` causes a warning
+   or error.
+
+As an alternative, one might think about using newtype structs for `c_long`:
+
+```rust
+#[cfg(unix)]
+struct c_long(i64);
+#[cfg(windows)]
+struct c_long(i32);
+
+impl FromPrimitive for c_long { ... }
+```
+
+This solution will enforce explicit conversions, however, they can't be used
+with `as` and create a lot of friction in the code:
+
+```rust
+fn gen(input: i64) -> i32 {
+    unsafe { Gen(FromPrimitive::from_i64(input).unwrap()).0 as i32 }
+}
+```
+
+(Note that the error checking performed by `FromPrimitive` above is not always
+desired.)
+
+Below we propose an alternative that sits between `type` and newtype structs.
+
+# Detailed design
+
+Rename `type` to `alias` to better document its function.
+
+Add a `newtype` keyword.
+
+## Syntax
+
+```rust
+newtype T = U;
+```
+
+where `T` and `U` are as in the definition of `alias`.
+
+(You might want to first skip the details and continue with the "Rationale"
+section below.)
+
+## Semantics
+
+`T` behaves as if it had been declared by `struct T(());` except as described
+below.
+
+### Representation
+
+`T` has the same representation in memory as `U` if any.
+
+### Casting
+
+Let `R1` be the binary, transitive, symmetric, irreflexive relation generated by
```

Comment:
If casts between types `newtype T = U` and built-in types `W != U` (both numeric) are not possible, then the system become completely useless because you can only cast between `c_long` and `i32` on Windows and only between `c_long` and `i64` on 64 bit unix.